services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/PushMonitorUtils.java

```diff
@@ -0,0 +1,112 @@
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class contains some common util methods for push monitoring purpose.
+ */
+public class PushMonitorUtils {
+  private static final Logger LOGGER = LogManager.getLogger(PushMonitorUtils.class);
+
+  /**
+   * This method checks Da Vinci client push status of all partitions from push status store and compute a final status.
+   * Inside each partition, this method will compute status based on all active Da Vinci instances.
+   * A Da Vinci instance sent heartbeat to controllers recently is considered active.
+   */
+  public static ExecutionStatusWithDetails getDaVinciPushStatusAndDetails(
+      PushStatusStoreReader reader,
+      String topicName,
+      int partitionCount,
+      Optional<String> incrementalPushVersion) {
+    if (reader == null) {
+      throw new VeniceException("PushStatusStoreReader is null");
+    }
+    LOGGER.info("Getting Da Vinci push status for topic: {}", topicName);
+    boolean allMiddleStatusReceived = true;
+    ExecutionStatus completeStatus = incrementalPushVersion.isPresent()
+        ? ExecutionStatus.END_OF_INCREMENTAL_PUSH_RECEIVED
+        : ExecutionStatus.COMPLETED;
+    ExecutionStatus middleStatus = incrementalPushVersion.isPresent()
+        ? ExecutionStatus.START_OF_INCREMENTAL_PUSH_RECEIVED
+        : ExecutionStatus.END_OF_PUSH_RECEIVED;
+    Optional<String> erroredInstance = Optional.empty();
+    String storeName = Version.parseStoreFromKafkaTopicName(topicName);
+    int completedPartitions = 0;
+    int totalInstanceCount = 0;
+    int liveInstanceCount = 0;
+    Set<Integer> incompletePartition = new HashSet<>();
+    for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+      Map<CharSequence, Integer> instances = reader.getPartitionStatus(
+          storeName,
+          Version.parseVersionFromVersionTopicName(topicName),
+          partitionId,
+          incrementalPushVersion);
+      boolean allInstancesCompleted = true;
+      totalInstanceCount += instances.size();
+      for (Map.Entry<CharSequence, Integer> entry: instances.entrySet()) {
+        ExecutionStatus status = ExecutionStatus.fromInt(entry.getValue());
+        boolean isInstanceAlive = reader.isInstanceAlive(storeName, entry.getKey().toString());
+        if (!isInstanceAlive) {
+          continue;
+        }
+        // We only compute status based on live instances.
+        liveInstanceCount++;
+        if (status == middleStatus) {
+          if (allInstancesCompleted) {
+            allInstancesCompleted = false;
+          }
+        } else if (status != completeStatus) {
+          if (allInstancesCompleted || allMiddleStatusReceived) {
+            allInstancesCompleted = false;
+            allMiddleStatusReceived = false;
+            if (status == ExecutionStatus.ERROR) {
```

Comment:
why the ERROR status detection needs to be in the `allInstancesCompleted || allMiddleStatusReceived` branch?

Reply:
I guess it is ok anywhere, we could move it out one layer I suppose from my current understanding.

Reply:
Refactored the whole block.

Reply:
I didn't pay attention that you just move this block of code around. Let't just keep it for now, the logic here is quite convoluted and we may want to clean it later.

Reply:
No worries, looks like everyone struggle to read this part so I tried to understand it myself and refactor a bit in the latest commit.